File: backend/internal/split/split.go
package split

import (
	"fmt"
	"sort"

	"github.com/DowLucas/chara/internal/money"
)

// MemberShare associates a split amount with a group member.
type MemberShare struct {
	MemberID string
	Share    money.Amount
}

// MemberPct associates a basis-point percentage (0–10000) with a group member.
type MemberPct struct {
	MemberID    string
	BasisPoints int // 10000 == 100%
}

// Equal splits total evenly across memberIDs. Remainder pennies are distributed
// to members in ascending lexicographic order of their ID.
func Equal(total money.Amount, memberIDs []string) ([]MemberShare, error) {
	if len(memberIDs) == 0 {
		return nil, fmt.Errorf("split: memberIDs must not be empty")
	}
	sorted := make([]string, len(memberIDs))
	copy(sorted, memberIDs)
	sort.Strings(sorted)

	parts := total.SplitEqual(len(sorted))
	result := make([]MemberShare, len(sorted))
	for i, id := range sorted {
		result[i] = MemberShare{MemberID: id, Share: parts[i]}
	}
	return result, nil
}

// Exact validates and returns the caller-supplied shares. Returns an error if
// any share is negative, any share exceeds money.MaxAmount, or the shares do
// not sum to total.
//
// The per-share bound is what makes the running sum safe: without it, two
// shares at the int64 ceiling wrap around to any target the caller likes, and
// a split that does not add up to the expense passes the sum check below.
func Exact(total money.Amount, shares []MemberShare) ([]MemberShare, error) {
	if total > money.MaxAmount {
		return nil, fmt.Errorf("split: total %s exceeds the maximum amount", total)
	}
	var sum money.Amount
	for _, s := range shares {
		if s.Share < 0 {
			return nil, fmt.Errorf("split: negative share for member %q", s.MemberID)
		}
		if s.Share > money.MaxAmount {
			return nil, fmt.Errorf("split: share for member %q exceeds the maximum amount", s.MemberID)
		}
		sum += s.Share
		if sum > money.MaxAmount {
			// Shares are non-negative, so the running sum only grows; once it
			// passes the cap it can never come back down to total.
			return nil, fmt.Errorf("split: shares sum above the maximum amount")
		}
	}
	if sum != total {
		return nil, fmt.Errorf("split: shares sum to %s, expected %s", sum, total)
	}
	return shares, nil
}

// Percentage splits total according to basis points (10000 == 100%). Returns an
// error if the basis points do not sum to 10000 or the slice is empty.
// Remainder pennies go to the member with the largest fractional remainder.
func Percentage(total money.Amount, pcts []MemberPct) ([]MemberShare, error) {
	if len(pcts) == 0 {
		return nil, fmt.Errorf("split: pcts must not be empty")
	}
	// total * basisPoints below must stay inside int64. money.MaxAmount is
	// chosen so that it does; anything larger is rejected rather than wrapped.
	if total < 0 || total > money.MaxAmount {
		return nil, fmt.Errorf("split: total %s is out of range", total)
	}
	var bpSum int
	for _, p := range pcts {
		bpSum += p.BasisPoints
	}
	if bpSum != 10000 {
		return nil, fmt.Errorf("split: basis points sum to %d, must be 10000", bpSum)
	}

	result := make([]MemberShare, len(pcts))
	var assigned money.Amount
	for i, p := range pcts {
		result[i] = MemberShare{
			MemberID: p.MemberID,
			Share:    money.Amount(int64(total) * int64(p.BasisPoints) / 10000),
		}
		assigned += result[i].Share
	}

	// Distribute any remaining pennies to members with the largest remainders.
	remainder := int(total - assigned)
	if remainder > 0 {
		type idx struct {
			i    int
			frac int64
		}
		fracs := make([]idx, len(pcts))
		for i, p := range pcts {
			fracs[i] = idx{i, int64(total)*int64(p.BasisPoints)%10000}
		}
		sort.Slice(fracs, func(a, b int) bool {
			return fracs[a].frac > fracs[b].frac
		})
		for i := 0; i < remainder; i++ {
			result[fracs[i].i].Share++
		}
	}

	return result, nil
}

File: app/lib/split.ts
/**
 * TypeScript port of `backend/internal/split/split.go`.
 *
 * IMPORTANT: This file must produce byte-identical output to the Go engine
 * for the same inputs. See `__tests__/split-fixture.test.ts` for the
 * cross-engine parity check.
 *
 * All amounts are int64 minor units, represented here as `bigint`.
 */

/** A computed share for one member, in minor units. */
export type Share = { memberId: string; amountMinor: bigint };

/** Input for the `exact` split method: each member's explicit share. */
export type ExactShareInput = { memberId: string; amountMinor: bigint };

/** Input for the `percentage` split method: each member's percentage (0–100). */
export type PercentageShareInput = { memberId: string; percentage: number };

/** Generic shape accepted by `computeSplits`. */
export type SplitShareInput = {
  memberId: string;
  amountMinor?: bigint;
  percentage?: number;
};

/**
 * Splits `total` evenly across `memberIds`. Remainder pennies are distributed
 * to members in ascending lexicographic order of their ID, matching the Go
 * implementation (`split.Equal` + `money.Amount.SplitEqual`).
 *
 * Output is sorted by memberId ascending.
 */
export function equalSplit(total: bigint, memberIds: string[]): Share[] {
  if (memberIds.length === 0) {
    throw new Error('split: memberIDs must not be empty');
  }
  const sorted = [...memberIds].sort();
  const n = BigInt(sorted.length);
  const base = total / n;
  // Remainder is positive when total > 0; for negative totals the Go behaviour
  // is dictated by Go's integer division truncating toward zero, which matches
  // JS BigInt division. The remainder may be negative for negative totals;
  // distribute its absolute count of pennies. But the Go code uses `int(a%n)`
  // and a loop `if i < remainder`, which silently skips when remainder<0 — to
  // stay byte-identical, mirror that semantics: only positive remainder buckets
  // get +1.
  const remainderBig = total - base * n;
  const remainder = Number(remainderBig); // small int, safe
  return sorted.map((id, i) => ({
    memberId: id,
    amountMinor: i < remainder ? base + 1n : base,
  }));
}

/**
 * Validates and returns caller-supplied shares unchanged (preserving input
 * order). Throws if any share is negative or the shares do not sum to `total`.
 * Matches `split.Exact`.
 */
export function exactSplit(total: bigint, shares: ExactShareInput[]): Share[] {
  let sum = 0n;
  for (const s of shares) {
    if (s.amountMinor < 0n) {
      throw new Error(`split: negative share for member "${s.memberId}"`);
    }
    sum += s.amountMinor;
  }
  if (sum !== total) {
    throw new Error(`split: shares sum to ${sum}, expected ${total}`);
  }
  return shares.map((s) => ({ memberId: s.memberId, amountMinor: s.amountMinor }));
}

/** Internal: percentage split keyed by basis points (10000 == 100%). */
type BasisPointShare = { memberId: string; basisPoints: number };

/**
 * Splits `total` according to basis points (10000 == 100%). Throws if basis
 * points do not sum to 10000 or the slice is empty. Remainder pennies go to
 * members with the largest fractional remainders. Matches `split.Percentage`.
 *
 * Note: input order is preserved (Go behaviour — `split.Percentage` does not
 * sort).
 */
export function percentageSplit(total: bigint, pcts: BasisPointShare[]): Share[] {
  if (pcts.length === 0) {
    throw new Error('split: pcts must not be empty');
  }
  let bpSum = 0;
  for (const p of pcts) {
    bpSum += p.basisPoints;
  }
  if (bpSum !== 10000) {
    throw new Error(`split: basis points sum to ${bpSum}, must be 10000`);
  }

  const result: Share[] = new Array(pcts.length);
  let assigned = 0n;
  const tenThousand = 10000n;
  for (let i = 0; i < pcts.length; i++) {
    const bp = BigInt(pcts[i].basisPoints);
    const share = (total * bp) / tenThousand; // truncation toward zero, matches Go
    result[i] = { memberId: pcts[i].memberId, amountMinor: share };
    assigned += share;
  }

  const remainder = Number(total - assigned);
  if (remainder > 0) {
    // Compute fractional remainders: total*bp % 10000 (Go uses int64, here bigint→number safe)
    const fracs = pcts.map((p, i) => ({
      i,
      frac: Number((total * BigInt(p.basisPoints)) % tenThousand),
    }));
    // Stable sort by frac descending. The Go implementation uses sort.Slice
    // which is NOT stable; to remain byte-identical for the same inputs we
    // mirror "sort by frac descending; ties keep original input order"
    // (which happens to be what most ties resolve to in practice with Go's
    // pdqsort on small slices, but the test fixture is the contract).
    fracs.sort((a, b) => b.frac - a.frac || a.i - b.i);
    for (let i = 0; i < remainder; i++) {
      result[fracs[i].i].amountMinor += 1n;
    }
  }

  return result;
}

/**
 * High-level split entry point used by the edit-expense screen. Routes to the
 * appropriate engine based on `method`.
 *
 * - `equal`: ignores `splits`; splits `amountMinor` across `participants`.
 * - `exact`: requires `splits[i].amountMinor` for every participant.
 * - `percentage`: requires `splits[i].percentage` for every participant
 *   (percentages are converted to basis points by multiplying by 100 and
 *   rounding to nearest).
 */
export function computeSplits(
  amountMinor: bigint,
  method: 'equal' | 'exact' | 'percentage',
  participants: string[],
  splits?: SplitShareInput[],
): Share[] {
  switch (method) {
    case 'equal':
      return equalSplit(amountMinor, participants);

    case 'exact': {
      if (!splits || splits.length === 0) {
        throw new Error('split: exact method requires splits');
      }
      const exact: ExactShareInput[] = splits.map((s) => {
        if (s.amountMinor === undefined) {
          throw new Error(`split: exact share for "${s.memberId}" is missing amountMinor`);
        }
        return { memberId: s.memberId, amountMinor: s.amountMinor };
      });
      return exactSplit(amountMinor, exact);
    }

    case 'percentage': {
      if (!splits || splits.length === 0) {
        throw new Error('split: percentage method requires splits');
      }
      const bps: BasisPointShare[] = splits.map((s) => {
        if (s.percentage === undefined) {
          throw new Error(`split: percentage share for "${s.memberId}" is missing percentage`);
        }
        // Convert percentage (0–100, possibly fractional) → basis points (0–10000).
        // Round to nearest to mirror UI behaviour; the validator below catches drift.
        const bp = Math.round(s.percentage * 100);
        return { memberId: s.memberId, basisPoints: bp };
      });
      return percentageSplit(amountMinor, bps);
    }

    default:
      throw new Error(`split: unknown method "${method}"`);
  }
}
